lib/src/rules/no_self_assign.dart

```diff
@@ -0,0 +1,150 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import '../analyzer.dart';
+
+const _desc = r"Don't assign a variable to itself";
+
+const _details = r'''
+**DON'T** assign a variable to itself. Usually this is a mistake.
+
+**BAD:**
+```dart
+class C {
+  int x;
+
+  C(int x) {
+    x = x;
+  }
+}
+```
+
+**GOOD:**
+```dart
+class C {
+  int x;
+
+  C(int x) : x = x;
+}
+```
+
+**GOOD:**
+```dart
+class C {
+  int x;
+
+  C(int x) {
+    this.x = x;
+  }
+}
+```
+
+**BAD:**
+```dart
+class C {
+  int _x = 5;
+
+  int get x => _x;
+
+  set x(int x) {
+    _x = x;
+    _customUpdateLogic();
+  }
+
+  void _customUpdateLogic() {
+    print('updated');
+  }
+
+  void example() {
+    x = x;
+  }
+}
+```
+
+**GOOD:**
+```dart
+class C {
+  int _x = 5;
+
+  int get x => _x;
+
+  set x(int x) {
+    _x = x;
+    _customUpdateLogic();
+  }
+
+  void _customUpdateLogic() {
+    print('updated');
+  }
+
+  void example() {
+    _customUpdateLogic();
+  }
+}
+```
+
+**BAD:**
+```dart
+class C {
+  int x = 5;
+
+  void update(C other) {
+    this.x = this.x;
+  }
+}
+```
+
+**GOOD:**
+```dart
+class C {
+  int x = 5;
+
+  void update(C other) {
+    this.x = other.x;
+  }
+}
+```
+
+''';
+
+class NoSelfAssign extends LintRule {
+  static const LintCode code = LintCode(
+      'no_self_assign', 'The variable or property is being assigned to itself.',
+      correctionMessage:
+          'Try removing the assignment that has no direct effect.');
+
+  NoSelfAssign()
+      : super(
+            name: 'no_self_assign',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  LintCode get lintCode => code;
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addAssignmentExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitAssignmentExpression(AssignmentExpression node) {
+    if (node.operator.type != TokenType.EQ) return;
```

Comment:
Is there any motivation to catch things like `??=` or `+=`?

Reply:
`+=`: no, because `a += a` is a perfectly valid way of multiplying something by 2 or something like that.

`??=`: haven't investigated how it looks